### ISSUE TITLE
Guidance for unsupported DT_RUNPATH

### ISSUE THREAD

**Issue by JonathonReinhart** [JonathonReinhart/staticx#188]
This issue is a placeholder for user guidance for #169 / #172.

**Comment by JonathonReinhart** [comment]
Please up-vote this comment if you are using GitHub Actions.

**Comment by jay0lee** [comment]
Upvoted. I actually self-compile Python 3.9.7, any suggestions on not using RPATH/RUNPATH in that case? I don't recall setting either as part of the configure/compile process for Python itself.

**Comment by jay0lee** [comment]
Here's a PoC Github Action I've setup to test this:

https://github.com/jay0lee/actions-hello-world/runs/3807920931

it seems to be some specific Ubuntu 18.04 issue as the same steps build a valid binary on Ubuntu 20.04.

**Comment by jay0lee** [comment]
Also FWIW I did test taking the StaticX binary build on Ubuntu 20.04 and running it on Debian 9 and CentOS 7. Both worked which is good enough for me. For now at least I'll just generate my legacy Linux static build on 20.04.

**Comment by JonathonReinhart** [comment]
Hi @jay0lee -- thanks for the feedback.

As I look at how to organize the data that comes in to this issue referenced from the code, can you do me a favor and open a new issue with your specific problems and findings? I'm not seeing anything in your logs related to this specific error, and your issue might be something else.

Also [this comment](https://github.com/JonathonReinhart/staticx/issues/188#issuecomment-933131870) was specifically trying to see data for how many people were running into this problem when using `actions/setup-python`. Since you're not -- you're building Python yourself -- would you mind dropping the thumbs-up as well?

Once the new issue is in place, I'll delete all of our comments here to keep this particular issue tidy.

Thanks!

**Comment by jay0lee** [comment]
Hi Jonathon, yes certainly. I opened #198 so you can clean this one up as needed. I suspect it's the same core issue with Ubuntu 18.04 but that's on me to prove. Sorry for the comment spam :-)

**Comment by BerndDasByte** [comment]
since staticx version v0.13.0 I get the following error message if I try to make a PyInstaller binary static:

`staticx: Unsupported PyInstaller input

One or more libraries included in the PyInstaller archive uses unsupported RPATH/RUNPATH tags:

  /tmp/staticx-pyi-gg2sp8i5/PyQt5/Qt/plugins/egldeviceintegrations/libqeglfs-emu-integration.so: DT_RUNPATH='$ORIGIN/../../lib'
  /tmp/staticx-pyi-gg2sp8i5/PyQt5/Qt/plugins/egldeviceintegrations/libqeglfs-kms-egldevice-integration.so: DT_RUNPATH='$ORIGIN/../../lib'
  /tmp/staticx-pyi-gg2sp8i5/PyQt5/Qt/plugins/egldeviceintegrations/libqeglfs-x11-integration.so: DT_RUNPATH='$ORIGIN/../../lib'
  /tmp/staticx-pyi-gg2sp8i5/PyQt5/Qt/plugins/iconengines/libqsvgicon.so: DT_RUNPATH='$ORIGIN/../../lib'
  /tmp/staticx-pyi-gg2sp8i5/PyQt5/Qt/plugins/imageformats/libqgif.so: DT_RUNPATH='$ORIGIN/../../lib'
  /tmp/staticx-pyi-gg2sp8i5/PyQt5/Qt/plugins/imageformats/libqicns.so: DT_RUNPATH='$ORIGIN/../../lib'
  /tmp/staticx-pyi-gg2sp8i5/PyQt5/Qt/plugins/imageformats/libqico.so: DT_RUNPATH='$ORIGIN/../../lib'
  /tmp/staticx-pyi-gg2sp8i5/PyQt5/Qt/plugins/imageformats/libqjpeg.so: DT_RUNPATH='$ORIGIN/../../lib'
  /tmp/staticx-pyi-gg2sp8i5/PyQt5/Qt/plugins/imageformats/libqsvg.so: DT_RUNPATH='$ORIGIN/../../lib'
  /tmp/staticx-pyi-gg2sp8i5/PyQt5/Qt/plugins/imageformats/libqtga.so: DT_RUNPATH='$ORIGIN/../../lib'
  ...`

With staticx versions **below** v0.13.0 I have no problems building from the **same,** PyInstaller binary ... What is the problem here?

**Comment by JonathonReinhart** [comment]
@BerndDasByte 

> With staticx versions **below** v0.13.0 I have no problems building from the **same,** PyInstaller binary ... What is the problem here?

The problem, as I believe has been thoroughly explained above, is that your PyInstalled binary includes libraries (In this case Qt5) which are linked with `RUNPATH`. Previously, staticx would have allowed this and produced executable, which when run, would potentially try to load libraries *from the target system*, which defeats the purpose (and guarantees) of StaticX. As described above, Staticx now tries to *fix* this problem by removing `RPATH`/`RUNPATH` where it can. However, if the problematic libraries are inside of a PyInstaller archive, then StaticX cannot easily modify (and thus does not modify) those.

The solution is not simple. You need to avoid libraries which link with `RUNPATH`. That may mean setting up a virtualenv and compiling them yourself. Unfortunately, I simply don't have the resources (i.e. time) right now to help much beyond that.

**Comment by Andrwe** [comment]
> Please up-vote this comment if you are seeing this new error message and you are using GitHub Actions, specifically [`actions/setup-python`](https://github.com/actions/setup-python).

Died for me in Github action run https://github.com/Andrwe/accointing-convert/runs/5000354491.

**Comment by joaompinto** [comment]
Short update, hopefully github will be able to work on the related issue https://github.com/actions/setup-python/issues/325 .
Meanwhile I will try to get some time to provide a GH action specially designed for the use staticx,

**Comment by th3w1zard1** [comment]
> RPATH
RPATH is allowed as long as the path is relative to $ORIGIN (the directory where the dynamic executable lives).
> 
> If staticx is complaining about your RPATH, it's probably because your library is referencing an absolute path like /usr/local/lib. If staticx were to allow this, then the application, when run on a foreign system, would try to load some dependent libraries from the target path, and not the staticx bundle.

I just don't understand... if the whole point is to statically link why not either A: copy the lib in question into the executable's directory or B: preferably, pack the lib into the executable binary.

> RUNPATH
> RUNPATH is always forbidden. This is because RUNPATH causes ld.so to completely disregard not only the RPATH set by staticx at program launch, for loading child dependencies of that library.
Same. Why is packing into the executable or moving into the executable's directory not an option?

I'm using pyinstaller to compile some scripts for cross-platform distribution. Half of those site-packages on linux use rpath/runpath, and after spending an hour trying to tweak the python installation, compiling from source, running patchelf, I'm not finding any solutions.
